pkg/protocol/engines.go

```diff
@@ -0,0 +1,261 @@
+
+	blockCache := blocks.New(evictionState, newStorage.Settings().APIProvider())
+	accountsManager := accountsledger.New(newStorage.Settings().APIProvider(), blockCache.Block, newStorage.AccountDiffs, newStorage.Accounts())
+
+	accountsManager.SetLatestCommittedSlot(latestCommitment.Slot())
+	if err = accountsManager.Rollback(slot); err != nil {
+		return nil, ierrors.Wrap(err, "failed to rollback accounts manager")
+	}
+
+	if err = evictionState.Rollback(newStorage.Settings().LatestFinalizedSlot(), slot); err != nil {
+		return nil, ierrors.Wrap(err, "failed to rollback eviction state")
+	}
+	if err = newStorage.Ledger().Rollback(slot); err != nil {
+		return nil, err
+	}
+
+	targetCommitment, err := newStorage.Commitments().Load(slot)
+	if err != nil {
+		return nil, ierrors.Wrapf(err, "error while retrieving commitment for target index %d", slot)
+	}
+
+	if err = newStorage.Settings().Rollback(targetCommitment); err != nil {
+		return nil, err
+	}
+
+	if err = newStorage.Rollback(slot); err != nil {
+		return nil, err
+	}
+
+	candidateEngine := e.loadEngineInstanceWithStorage(newEngineAlias, newStorage)
+
+	// rollback attestations already on created engine instance, because this action modifies the in-memory storage.
+	if err = candidateEngine.Attestations.Rollback(slot); err != nil {
+		return nil, ierrors.Wrap(err, "error while rolling back attestations storage on candidate engine")
+	}
+
+	return candidateEngine, nil
+}
+
+// CleanupCandidates removes all engine instances that are not the main engine.
+func (e *Engines) CleanupCandidates() error {
+	activeDir := filepath.Base(e.Main.Get().Storage.Directory())
+
+	dirs, err := e.directory.SubDirs()
+	if err != nil {
+		return ierrors.Wrapf(err, "unable to list subdirectories of %s", e.directory.Path())
+	}
+	for _, dir := range dirs {
+		if dir == activeDir {
+			continue
+		}
+		if err := e.directory.RemoveSubdir(dir); err != nil {
+			return ierrors.Wrapf(err, "unable to remove subdirectory %s", dir)
+		}
+	}
+
+	return nil
+}
+
+// infoFilePath returns the path to the engine info file.
+func (e *Engines) infoFilePath() string {
+	return e.directory.Path(engineInfoFile)
+}
+
+// loadEngineInstanceFromSnapshot loads an engine instance from a snapshot.
+func (e *Engines) loadEngineInstanceFromSnapshot(engineAlias string, snapshotPath string) *engine.Engine {
+	errorHandler := func(err error) {
+		e.protocol.LogError("engine error", "err", err, "name", engineAlias[0:8])
+	}
+
+	e.protocol.Options.EngineOptions = append(e.protocol.Options.EngineOptions, engine.WithSnapshotPath(snapshotPath))
+
+	return e.loadEngineInstanceWithStorage(engineAlias, storage.Create(e.directory.Path(engineAlias), DatabaseVersion, errorHandler, e.protocol.Options.StorageOptions...))
+}
+
+// loadEngineInstanceWithStorage loads an engine instance with the given storage.
+func (e *Engines) loadEngineInstanceWithStorage(engineAlias string, storage *storage.Storage) *engine.Engine {
```

Comment:
please split up in multiple lines. it's unreadable and hard to edit if the line is 50m long and you need to scroll to the right